File: 47-permutations-ii/47-permutations-ii.go

func permuteUnique(nums []int) [][]int {
    n := len(nums)
    var (
        res [][]int
        dfs func(int)
    )
    
    dfs = func(idx int) {
        if idx == n {
            tmp := make([]int, n)
            copy(tmp, nums)
            res = append(res, tmp)
        } 
        
        set := make(map[int]struct{})
        for i := idx; i < n; i++ {
            if _, exists := set[nums[i]]; !exists {
                nums[i], nums[idx] = nums[idx], nums[i]
                dfs(idx+1)
                nums[i], nums[idx] = nums[idx], nums[i]
                set[nums[i]] = struct{}{}
            }
        }
    }
    
    dfs(0)
    return res
}

File: 47-permutations-ii/47-permutations-ii.py
# 47. Permutations II
# time: O(n*n!)
# space: O(n*n!)

from collections import defaultdict

class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        res = []
        n = len(nums)
        freqMap = defaultdict(int)
        for num in nums:
            freqMap[num] += 1
        def helper(perm: List[int]):
            nonlocal res
            nonlocal freqMap
            if len(perm) == n:
                res.append(perm[:])
                return
            for num_key, cnt_val in freqMap.items():
                if cnt_val == 0:
                    continue
                perm.append(num_key)
                freqMap[num_key] -= 1
                helper(perm)
                freqMap[num_key] += 1
                perm.pop()
        
        helper([])
        return res